### PR TITLE
Track numpy/cython versions used in build

### PR DESCRIPTION
**Pull request by aragilar**
This complements #1619 (in that we'll be able to see which versions were used within the PEP517/8 build), and hopefully will make debugging issues like #1538 easier.

### CODE REVIEW

**Comment by takluyver** [comment]
Nice! Here's the output in one of the CI jobs:

```
Summary of the h5py configuration
---------------------------------

h5py    2.10.0
HDF5    1.10.5
Python  3.7.8 (default, Jun 29 2020, 08:44:51) 
[GCC 5.4.0 20160609]
sys.platform    linux
sys.maxsize     9223372036854775807
numpy   1.19.1
cython (built against) 0.29.21
numpy (built against) 1.19.1
HDF5 (built against) 1.10.5
```

File: h5py/h5.pyx [hunk 1]
```diff
@@ -21,6 +21,8 @@ INDEX_NAME      = H5_INDEX_NAME       # Index on names
 INDEX_CRT_ORDER = H5_INDEX_CRT_ORDER  # Index on creation order
 
 HDF5_VERSION_COMPILED_AGAINST = HDF5_VERSION
+NUMPY_VERSION_COMPILED_AGAINST = NUMPY_BUILD_VERSION
+CYTHON_VERSION_COMPILED_AGAINST = CYTHON_BUILD_VERSION
```
Comment (takluyver) [inline h5py/h5.pyx:25]:
```suggestion
CYTHON_VERSION_COMPILED_WITH = CYTHON_BUILD_VERSION
```

Small nitpick: I think 'compiled against' means something exposes a C API which we use; Cython is a bit different, in that it's used as part of our build process but we don't end up linked against libcython or something.

File: h5py/version.py [hunk 2]
```diff
@@ -49,9 +49,17 @@
 sys.platform    %(platform)s
 sys.maxsize     %(maxsize)s
 numpy   %(numpy)s
-""" % { 'h5py': version,
-        'hdf5': hdf5_version,
-        'python': sys.version,
-        'platform': sys.platform,
-        'maxsize': sys.maxsize,
-        'numpy': numpy.__version__ }
+cython (built against) %(cython_version)s
```
Comment (takluyver) [inline h5py/version.py:52]:
```suggestion
cython (built with) %(cython_version)s
```

File: h5py/version.py [hunk 3]
```diff
@@ -49,9 +49,17 @@
...
+numpy (built against) %(numpy_build_version)s
+HDF5 (built against) %(hdf5_build_version)s
+""" % {
+    'h5py': version,
+    'hdf5': hdf5_version,
+    'python': sys.version,
+    'platform': sys.platform,
+    'maxsize': sys.maxsize,
+    'numpy': numpy.__version__,
+    'cython_version': _h5.CYTHON_VERSION_COMPILED_AGAINST,
```
Comment (takluyver) [inline h5py/version.py:62]:
```suggestion
    'cython_version': _h5.CYTHON_VERSION_COMPILED_WITH,
```

File: h5py/h5.pyx (hunk 1 above)
Comment (aragilar) [reply h5py/h5.pyx]: Done (definitely reads better).

**Pull request merged**